test: fix use of Float/Rational in tests

oscarbenjamin opened the pull request:
This commit changes tests that assert comparisons between Floats and
Rationals as well as Floats with different precision. Where a tested
function returns a Float (or an expression involving Floats) then the
test should compare the returned value with a Float. Likewise if exact
Rationals are returned then the test should assert that.

There were also many tests that use Floats with specific precision but
compare against Floats with different precision. This commit ensures
that a correct precision Float is always used in these comparisons.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

This extracts some of the harmless changes from #20033.

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
This is ready for review.

ThePauliPrinciple commented on sympy/core/tests/test_numbers.py:
```diff
@@ -2002,7 +2002,7 @@ def test_issue_10020():
 def test_invert_numbers():
     assert S(2).invert(5) == 3
     assert S(2).invert(Rational(5, 2)) == S.Half
-    assert S(2).invert(5.) == 0.5
```
Is `5.1` not a float here? Maybe because of sympify/python? It appears to me as if this test was meant to return a float, but perhaps it stopped doing so during some commit?

oscarbenjamin replied:
I'm not sure if this is something that was changed at some point or if the original author simply didn't check.

oscarbenjamin replied:
The code that handles this is here:
https://github.com/sympy/sympy/blob/857a97e66c6bb162eb922179aea20226c7a0ed3b/sympy/core/numbers.py#L546-L550
I have no idea what it is trying to do but it ignores the type of `m` if `m > 1`. Hence:
```python
In [1]: S(2).invert(5.)
Out[1]: 1/2

In [2]: S(2.).invert(5.)
Out[2]: 0.500000000000000

In [3]: S(2.).invert(5)
Out[3]: 0.500000000000000
```
If both are integers:
```python
In [4]: S(2).invert(5)
Out[4]: 3
```

ThePauliPrinciple commented on sympy/core/tests/test_sympify.py:
```diff
@@ -626,10 +626,11 @@ def equal(x, y):
     assert equal(sympify(np.complex64(1 + 2j)), S(1.0 + 2.0*I))
     assert equal(sympify(np.complex128(1 + 2j)), S(1.0 + 2.0*I))
-    assert equal(sympify(np.longcomplex(1 + 2j)), S(1.0 + 2.0*I))
+    assert equal(sympify(np.longcomplex(1 + 2j)),
```
Why does this become `precision=64`? Shouldn't it be 128 since a longcomplex is 256 bit https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.clongdouble

oscarbenjamin replied:
The precision refers to the number of bits in the mantissa so e.g. np.float64 corresponds to a precision of 53 bits. I think np.float128 should have 113 bits of precision. That's not what sympify does though:
```python
In [29]: sympify(np.longcomplex(1 + 2j)).args[0]._prec
Out[29]: 64
```

ThePauliPrinciple replied:
So then one could state that the result of `sympify(np.longcomplex(1 + 2j))` is incorrect? Since it creates a `Float` with the wrong precision?

asmeurer replied:
The code in `sympify` is using `np.finfo(dtype).nmant`, which does return `63` for `np.longcomplex` https://github.com/sympy/sympy/blob/42248dd7be300bc3893642fd92793c0ff4c19a89/sympy/core/sympify.py#L91

oscarbenjamin replied:
It comes from here:
https://github.com/sympy/sympy/blob/42248dd7be300bc3893642fd92793c0ff4c19a89/sympy/core/sympify.py#L90-L96
We can try that manually:
```python
In [7]: np.finfo(np.float128(1)).nmant + 1
Out[7]: 64

In [8]: np.finfo(np.float64(1)).nmant + 1
Out[8]: 53
```
The numpy docs say 
```
nmant : int
The number of bits in the mantissa.
```
https://numpy.org/doc/stable/reference/generated/numpy.finfo.html

However if you look here then you can see that the actual size of these datatypes is platform dependent:
https://numpy.org/doc/stable/user/basics.types.html#extended-precision

So actually np.float128 is most likely not the same as IEEE 128 bit floating point. I guess that on my computer (and also in CI) what we actually get is an 80-bit floating point number (as was used in x87 FPUs in the 32 bit days):
https://en.wikipedia.org/wiki/Extended_precision#Extended_precision_implementations

ThePauliPrinciple replied:
I wonder if `finfo` then is not accurate. Why would `sympy.sympify(np.complex64(1 + 2j)).args[0]._prec == sympy.sympify(np.complex128(1 + 2j)).args[0]._prec == 53`?

ThePauliPrinciple replied:
I see, so then the test should probably be rewritten in some platform independent way?

asmeurer replied:
See https://stackoverflow.com/questions/9062562/what-is-the-internal-precision-of-numpy-float128/17023995#17023995 and https://github.com/numpy/numpy/issues/10288.

oscarbenjamin replied:
I've changed the test to be based on np.finfo

ThePauliPrinciple commented on sympy/physics/units/tests/test_quantities.py:
```diff
@@ -253,7 +253,7 @@ def test_mul_div():
 
 def test_units():
-    assert convert_to((5*m/s * day) / km, 1) == 432
+    assert convert_to((5*m/s * day) / km, 1) == 432.0
```
Is this not exactly 432? Why should this be a float?

oscarbenjamin replied:
I'm not sure where it comes from.

smichr replied:
```diff
diff --git a/sympy/physics/units/util.py b/sympy/physics/units/util.py
index 08f611858e..ee6e4c6ee5 100644
--- a/sympy/physics/units/util.py
+++ b/sympy/physics/units/util.py
@@ -99,16 +99,20 @@ def convert_to(expr, target_units, unit_system="SI"):
         target_units = [target_units]
 
     if isinstance(expr, Add):
-        return Add.fromiter(convert_to(i, target_units, unit_system) for i in expr.args)
+        return Add.fromiter(convert_to(i, target_units, unit_system)
+            for i in expr.args)
 
     expr = sympify(expr)
+    target_units = sympify(target_units)
 
     if not isinstance(expr, Quantity) and expr.has(Quantity):
-        expr = expr.replace(lambda x: isinstance(x, Quantity), lambda x: x.convert_to(target_units, unit_system))
+        expr = expr.replace(lambda x: isinstance(x, Quantity),
+            lambda x: x.convert_to(target_units, unit_system))
 
     def get_total_scale_factor(expr):
         if isinstance(expr, Mul):
-            return reduce(lambda x, y: x * y, [get_total_scale_factor(i) for i in expr.args])
+            return reduce(lambda x, y: x * y,
+                [get_total_scale_factor(i) for i in expr.args])
         elif isinstance(expr, Pow):
             return get_total_scale_factor(expr.base) ** expr.exp
         elif isinstance(expr, Quantity):
@@ -120,7 +124,9 @@ def get_total_scale_factor(expr):
         return expr
 
     expr_scale_factor = get_total_scale_factor(expr)
-    return expr_scale_factor * Mul.fromiter((1/get_total_scale_factor(u) * u) ** p for u, p in zip(target_units, depmat))
+    return expr_scale_factor * Mul.fromiter(
+        (1/get_total_scale_factor(u)*u)**p for u, p in
+        zip(target_units, depmat))
 
 
 def quantity_simplify(expr, across_dimensions: bool=False, unit_system=None):
```

The `target_units` needs to be sympified.

ThePauliPrinciple commented on sympy/series/tests/test_limits.py:
```diff
@@ -372,10 +372,10 @@ def test_issue_4547():
 def test_issue_5164():
     assert limit(x**0.5, x, oo) == oo**0.5 is oo
-    assert limit(x**0.5, x, 16) == S(16)**0.5
+    assert limit(x**0.5, x, 16) == 4 # Should this be a float?
```
I believe so, why would it lose its "floatness"?

oscarbenjamin replied:
Probably because the float is converted into (or interpreted as) a Rational somewhere.

oscarbenjamin replied:
I reverted this change

ThePauliPrinciple commented:
Structural comparisons (`==`) between floats of different precision should clearly result in unequal, but it is (still) not so clear to me what should happen for `Eq`. 

In general I feel that float inputs into functions should be remembered. If people feel that in their case floats can safely be converted to integers they can round it explicitly, it should not be assumed that it can be.

oscarbenjamin commented:
To be clear the purpose of this PR is not to change any behaviour of SymPy. The idea is just to make sure that the tests are explicit about what they are testing. I also want to reduce the diff for #20033 so that it's clearer what the actual changes are and so that I can also use it to extract other more useful changes.

ThePauliPrinciple commented:
> To be clear the purpose of this PR is not to change any behaviour of SymPy. The idea is just to make sure that the tests are explicit about what they are testing. I also want to reduce the diff for #20033 so that it's clearer what the actual changes are and so that I can also use it to extract other more useful changes.

This is how I interpreted this PR. My concern is mostly that some tests have now been changed to, what appears to me, as an incorrect result, when float comparisons should be considered more strictly. e.g. `assert limit(x**0.5, x, 16) == 4` to me should be `4.0 == 4`, which should be false. 

So while this paves the way to a better comparison of floats (and this is absolutely the right direction), it might also be hiding incorrect results of other tests because they are based on the current outcome of the tests and not on what the outcome should be. These tests are currently (as in, before this PR) only passing because float comparisons are implemented in an inconsistent way in the first place.

oscarbenjamin commented:
> My concern is mostly that some tests have now been changed to, what appears to me, as an incorrect result, when float comparisons should be considered more strictly.

I have added comments in places where I thought that was the case. It is always necessary to apply some interpretation when reading the results of the test suite.

> it might also be hiding incorrect results of other tests because they are based on the current outcome of the tests and not on what the outcome should be.

I don't know what you mean by this. This PR just makes it explicit in the test code itself what is being returned. If we want to enforce that in future then we can use something like #23499 except only have it run in CI rather than actually making the change for end users. In the mean time #23499 can be used to see whether there are any tests that mix floats and rationals in comparisons.

ThePauliPrinciple commented:
> I don't know what you mean by this

Again, the example `assert limit(x**0.5, x, 16) == 4`. The argument here is apparently to make the change from ` ... == 4.0` to `... == 4` because that is currently the result of `limit(x**0.5, x, 16)` is `4`. And you want to be able to later on change the structural comparison of `Float` and `Integer` to fail.

While I understand this reasoning, the result is now that we are testing for `limit(x**0.4, x, 16) == 4`, which I think is not the right test. I would argue that this should return `False`, eventhough it currently does not.

oscarbenjamin commented:
> While I understand this reasoning, the result is now that we are testing for `limit(x**0.4, x, 16) == 4`, which I think is not the right test.

That's why I added a comment on that line "should this be a float?". The point is to make it clear to someone looking at it that the test itself can be questioned and should not be considered definitive. The comment will also show up in the output of the test runner:
```
Traceback (most recent call last):
  File "/home/oscar/current/sympy/sympy.git/sympy/series/tests/test_limits.py", line 375, in test_issue_5164
    assert limit(x**0.5, x, 16) == 4 # Should this be a float?
AssertionError
```

ThePauliPrinciple commented:
The original test was correct, so I don't think it should be replaced with the incorrect result and a comment.

oscarbenjamin commented:
> The original test was correct, so I don't think it should be replaced with the incorrect result and a comment.

I will revert the change but I think that if the goal is not to hide potentially incorrect behaviour then leaving the test as it is in master is worse.

ThePauliPrinciple commented:
> if the goal is not to hide potentially incorrect behaviour then leaving the test as it is in master is worse.

This is what I don't understand. How I see it, when the change is made for `Float` to compare inequal to `Integer` this test will start to fail. It can then become an "xfail" test to indicated it is known to fail but should eventually pass. Or alternatively, in the mean time `limits` is changed to give `4.0` as a result.

oscarbenjamin commented:
> when the change is made for `Float` to compare inequal to `Integer`

I don't think it's a foregone conclusion that that change would ever be made (there isn't really any agreement if you look at the PR). The fact that I am working on this now is not because I think that the change is necessarily going to happen but rather because I want to extract other useful things from the PR.

> It can then become an "xfail" test to indicated it is known to fail but should eventually pass.

In practice xfail tests are pretty useless in sympy. Most contributors don't notice them because they don't cause CI to fail. Most people who write xfail tests don't know how to write a test that actually works so that it will pass one day.

To me the most basic point in the test code is that I want to be able to judge from reading the code in the test what the current behaviour of sympy is. If there is some behaviour that I care about then I want the test suite to show me when that behaviour changes (for better or for worse).

Currently in master those tests are misleading: they appear to assert a result that is not what actually happens (and then still pass). That means that the test code itself gives a misleading impression of current sympy behaviour. Whether that behaviour is correct or should be changed is a separate point and this is where I think that there should be more comments in the test code.

ThePauliPrinciple commented:
In previous versions this limit actually did return a float:
```python
>>> import sympy
>>> x = sympy.abc.x
>>> print(sympy.limit(x**0.5, x, 16))
>>> print(sympy.__version__)
4.00000000000000
1.7.1
```
So we're really looking at a regression here that wasn't picked up because of how `Float` currently compares.

oscarbenjamin commented:
I've bisected the change to limit to cec5c2b7a79a37d1fcf181eeca1a4772d7d8ca3c from #22213. The change seems deliberate to me

CC @0sidharth

ThePauliPrinciple commented:
Alright, if these changes are intentional than that is what it is. It would be nice if we could have some consistent definition of what a `Float` is in SymPy and when it is allowed to be automatically converted to some other type without the user knowing about it.

0sidharth commented:
These changes are indeed intended. I don't think a better alternative for dealing with the original issue of #22213 exists, so maybe just mentioning to the user when this conversion is done on purpose would be the way to go.

oscarbenjamin commented:
Maybe the conversion to rational could be temporary and then the result could be converted back to float later. You can something similar here:
https://github.com/sympy/sympy/blob/832c24fec1046eaa544a4cab4c69e3af3e651759/sympy/solvers/solveset.py#L3520-L3525
Then later:
https://github.com/sympy/sympy/blob/832c24fec1046eaa544a4cab4c69e3af3e651759/sympy/solvers/solveset.py#L3550-L3553

0sidharth commented:
Thanks for the suggestion, this should be easily implementable. Although I don't like having to add such a conditional check, the consistency is probably valuable, so I'll try to see if I can get this in within the next day.

oscarbenjamin committed:
test: fix use of Float/Rational in tests

This commit changes tests that assert comparisons between Floats and
Rationals as well as Floats with different precision. Where a tested
function returns a Float (or an expression involving Floats) then the
test should compare the returned value with a Float. Likewise if exact
Rationals are returned then the test should assert that.

There were also many tests that use Floats with specific precision but
compare against Floats with different precision. This commit ensures
that a correct precision Float is always used in these comparisons.

oscarbenjamin commented:
I've reverted the changes to units tests after #23537

smichr commented:
Should this wait for #23539?

oscarbenjamin commented:
> Should this wait for #23539?

No, I've reverted the change to that test.

There shouldn't be anything controversial here. Reviews welcome.

asmeurer commented on sympy/codegen/ast.py:
```diff
@@ -1535,7 +1535,7 @@ def as_Declaration(self, **kwargs):
         True
         >>> decl2 = x.as_Declaration(value=42.0)
-        >>> decl2.variable.value == 42
+        >>> decl2.variable.value == 42.0
```
I guess you don't have to fix this here, but I don't like doctests that are written like this. I would rather see

```py
>>> decl2.variable.value
42.0
```

We should only write doctests with `==` this when it's impossible to doctest the output for some reason (and even then, maybe just skipping the test can be better). 

Even when the output is `None` (which doctest hides), I'd rather see

```py
>>> print(x.is_integer)
None
```

than

```py
>>> x.is_integer == None
True
```

If these doctests were written this way in the first place you wouldn't need to fix them here.

asmeurer commented:
Just scrolling through this looks good, and does make #20033 convincing, since quite a few of these tests were basically wrong before. 

I'm wondering in #20033, how do Floats with the non-default precision work when comparing to `float`? Will `Float('1.0', 5) == 1.0` give True or False?

oscarbenjamin commented:
> Will `Float('1.0', 5) == 1.0` give True or False?

In #20033 a normal Python float is considered to be whatever sympify would make of it:
```python
In [3]: srepr(S(1.0))
Out[3]: "Float('1.0', precision=53)"

In [4]: Float('1.0', 5) == 1.0
Out[4]: False

In [5]: S(1.0) == 1.0
Out[5]: True
```
When I can get this merged I'll update #20033 to see what can be done with it.

oscarbenjamin commented:
Okay I'll merge this now.
